pkg/controller/registry/resolver/predicates.go

```diff
@@ -0,0 +1,376 @@
+		}
+		var prop opregistry.GVKProperty
+		err := json.Unmarshal([]byte(p.Value), &prop)
+		if err != nil {
+			continue
+		}
+		if prop.Kind == g.api.Kind && prop.Version == g.api.Version && prop.Group == g.api.Group {
+			return true
+		}
+	}
+	return false
+}
+
+func (g gvkPredicate) String() string {
+	return fmt.Sprintf("with group: %s, version: %s, kind: %s", g.api.Group, g.api.Version, g.api.Kind)
+}
+
+type skipRangeIncludesPredication struct {
+	version semver.Version
+}
+
+func SkipRangeIncludesPredicate(version semver.Version) OperatorPredicate {
+	return skipRangeIncludesPredication{version: version}
+}
+
+func (s skipRangeIncludesPredication) Test(o *Operator) bool {
+	semverRange, err := o.SemverRange()
+	return err == nil && semverRange(s.version)
+}
+
+func (s skipRangeIncludesPredication) String() string {
+	return fmt.Sprintf("skip range includes: %v", s.version.String())
+}
+
+type replacesPredicate string
+
+func ReplacesPredicate(replaces string) OperatorPredicate {
+	return replacesPredicate(replaces)
+}
+
+func (r replacesPredicate) Test(o *Operator) bool {
+	if o.Replaces() == string(r) {
+		return true
+	}
+	for _, s := range o.Skips() {
+		if s == string(r) {
+			return true
+		}
+	}
+	return false
+}
+
+func (r replacesPredicate) String() string {
+	return fmt.Sprintf("replaces: %v", string(r))
+}
+
+type andPredicate struct {
+	predicates []OperatorPredicate
+}
+
+func And(p ...OperatorPredicate) OperatorPredicate {
+	return andPredicate{
+		predicates: p,
+	}
+}
+
+func (p andPredicate) Test(o *Operator) bool {
+	for _, predicate := range p.predicates {
+		if predicate.Test(o) == false {
+			return false
+		}
+	}
+	return true
+}
+
+func (p andPredicate) String() string {
+	var b bytes.Buffer
+	for i, predicate := range p.predicates {
+		b.WriteString(predicate.String())
+		if i != len(p.predicates)-1 {
+			b.WriteString(" and ")
+		}
+	}
+	return b.String()
+}
+
+func Or(p ...OperatorPredicate) OperatorPredicate {
+	return orPredicate{
+		predicates: p,
+	}
+}
+
+type orPredicate struct {
+	predicates []OperatorPredicate
+}
+
+func (p orPredicate) Test(o *Operator) bool {
+	for _, predicate := range p.predicates {
+		if predicate.Test(o) == true {
+			return true
+		}
+	}
+	return false
+}
+
+func (p orPredicate) String() string {
+	var b bytes.Buffer
+	for i, predicate := range p.predicates {
+		b.WriteString(predicate.String())
+		if i != len(p.predicates)-1 {
+			b.WriteString(" or ")
+		}
+	}
+	return b.String()
+}
+
+type OperatorPredicateFunc func(*Operator) bool
+
+func (opf OperatorPredicateFunc) Test(o *Operator) bool {
+	return opf(o)
+}
+
+func (opf OperatorPredicateFunc) String() string {
+	return ""
+}
+
+func True() OperatorPredicate {
+	return OperatorPredicateFunc(func(*Operator) bool {
+		return true
+	})
+}
+
+func False() OperatorPredicate {
+	return OperatorPredicateFunc(func(*Operator) bool {
+		return false
+	})
+}
+
+func CountingPredicate(p OperatorPredicate, n *int) OperatorPredicate {
```

Comment:
If you decide to remove `OperatorPredicateFunc`, I'd expect the String() implementation for a counting predicate to delegate to the predicate that it wraps -- the counter doesn't affect the meaning of the predicate.

Reply:
I'd explored this too, but at the heart of the counting predicate it's calling p.Test() for the predicate it receives, which requires the `resolver.Operator` as it's argument, which is not available statically, i.e the `OperatorPredicate` func type that's implemented within this method does make sense.